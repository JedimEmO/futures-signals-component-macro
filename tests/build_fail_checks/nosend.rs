use futures_signals::signal::Signal;
use futures_signals_component_macro::component;

#[component(render_fn = render_send)]
struct NeedsSend<T: Send = (), TNotSend: Clone = ()> {
    #[signal]
    send_me: T,

    #[signal]
    don_not_send_me: TNotSend,
}

fn render_send(props: impl NeedsSendPropsTrait + 'static) -> i32 {
    let NeedsSendProps {
        send_me,
        don_not_send_me
    } = props.take();

    consume_send(don_not_send_me.unwrap());

    42
}

fn consume_send(_: impl Signal<Item=impl Send>) {}

fn main() {}