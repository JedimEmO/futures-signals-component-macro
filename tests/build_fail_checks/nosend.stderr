error[E0277]: `<impl NeedsSendPropsTrait + 'static as NeedsSendPropsTrait>::TNotSend` cannot be sent between threads safely
  --> tests/build_fail_checks/nosend.rs:19:5
   |
19 |     consume_send(don_not_send_me.unwrap());
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `<impl NeedsSendPropsTrait + 'static as NeedsSendPropsTrait>::TNotSend` cannot be sent between threads safely
   |
   = help: the trait `Send` is not implemented for `<impl NeedsSendPropsTrait + 'static as NeedsSendPropsTrait>::TNotSend`
note: required by a bound in `consume_send`
  --> tests/build_fail_checks/nosend.rs:24:42
   |
24 | fn consume_send(_: impl Signal<Item=impl Send>) {}
   |                                          ^^^^ required by this bound in `consume_send`
help: introduce a type parameter with a trait bound instead of using `impl Trait`
   |
13 | fn render_send<N: NeedsSendPropsTrait + 'static>(props: N) -> i32 where <N as NeedsSendPropsTrait>::TNotSend: Send {
   |               ++++++++++++++++++++++++++++++++++        ~         ++++++++++++++++++++++++++++++++++++++++++++++++
